File: clients/go/apex_kv.go
// Package apexkv provides a Go client for the APEX-KV distributed key-value store.
//
// Usage:
//
//	client, err := apexkv.New("127.0.0.1:7001")
//	if err != nil { log.Fatal(err) }
//	defer client.Close()
//
//	if err := client.Put("hello", "world"); err != nil { log.Fatal(err) }
//	val, err := client.Get("hello")   // "world", nil
//	err = client.Del("hello")
package apexkv

import (
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"net"
	"sync"
	"sync/atomic"
)

// Wire protocol constants (must match apex/protocol.hpp)
const (
	wireMagic   uint32 = 0xA9ECBD10
	hdrSize            = 16
)

type cmd uint8

const (
	cmdGET          cmd = 0x01
	cmdPUT          cmd = 0x02
	cmdDEL          cmd = 0x03
	cmdPING         cmd = 0x04
	cmdMETRICS      cmd = 0x05
	cmdOK           cmd = 0x10
	cmdVALUE        cmd = 0x11
	cmdNOTFOUND     cmd = 0x12
	cmdERROR        cmd = 0x13
	cmdREDIRECT     cmd = 0x14
	cmdPONG         cmd = 0x15
	cmdMETRICSRESP  cmd = 0x16
)

// Sentinel errors
var (
	ErrNotFound = errors.New("key not found")
	ErrNotLeader = errors.New("not leader")
)

// NotLeaderError carries the redirect address.
type NotLeaderError struct{ Addr string }
func (e *NotLeaderError) Error() string { return "not leader, redirect to " + e.Addr }

// Metrics holds server-side statistics returned by the Metrics() call.
type Metrics struct {
	OpsGet     uint64
	OpsPut     uint64
	P99GetUs   uint64
	P99PutUs   uint64
}

// Client is a thread-safe APEX-KV client.
type Client struct {
	mu   sync.Mutex
	conn net.Conn
	seq  atomic.Uint32
	addr string
}

// New creates a connected client. addr is "host:port".
func New(addr string) (*Client, error) {
	c := &Client{addr: addr}
	if err := c.reconnect(); err != nil {
		return nil, fmt.Errorf("apexkv: connect %s: %w", addr, err)
	}
	return c, nil
}

func (c *Client) reconnect() error {
	conn, err := net.Dial("tcp", c.addr)
	if err != nil {
		return err
	}
	if tc, ok := conn.(*net.TCPConn); ok {
		_ = tc.SetNoDelay(true)
	}
	c.conn = conn
	return nil
}

// Close shuts down the connection.
func (c *Client) Close() error {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.conn != nil {
		return c.conn.Close()
	}
	return nil
}

// Ping returns true if the node is alive.
func (c *Client) Ping() (bool, error) {
	rc, _, err := c.rpc(cmdPING, nil, true)
	return rc == cmdPONG, err
}

// Get returns the value for key, or ErrNotFound if not present.
func (c *Client) Get(key string) (string, error) {
	payload := encodeStr(key)
	rc, resp, err := c.rpc(cmdGET, payload, false)
	if err != nil {
		return "", err
	}
	switch rc {
	case cmdVALUE:
		return decodeStr(resp), nil
	case cmdNOTFOUND:
		return "", ErrNotFound
	case cmdERROR:
		return "", fmt.Errorf("apexkv: server error: %s", decodeStr(resp))
	default:
		return "", fmt.Errorf("apexkv: unexpected response 0x%02x", rc)
	}
}

// Put writes key=value. Follows leader redirects automatically.
func (c *Client) Put(key, value string) error {
	payload := append(encodeStr(key), encodeStr(value)...)
	rc, resp, err := c.rpc(cmdPUT, payload, true)
	if err != nil {
		return err
	}
	if rc == cmdOK {
		return nil
	}
	return fmt.Errorf("apexkv: put failed (%02x): %s", rc, decodeStr(resp))
}

// Del deletes a key. Returns nil whether or not the key existed.
func (c *Client) Del(key string) error {
	payload := encodeStr(key)
	_, _, err := c.rpc(cmdDEL, payload, true)
	return err
}

// GetOrDefault returns the value or a default string if not found.
func (c *Client) GetOrDefault(key, def string) (string, error) {
	v, err := c.Get(key)
	if errors.Is(err, ErrNotFound) {
		return def, nil
	}
	return v, err
}

// Metrics fetches server-side statistics.
func (c *Client) Metrics() (Metrics, error) {
	rc, payload, err := c.rpc(cmdMETRICS, nil, false)
	if err != nil {
		return Metrics{}, err
	}
	if rc != cmdMETRICSRESP || len(payload) < 32 {
		return Metrics{}, fmt.Errorf("apexkv: bad metrics response")
	}
	return Metrics{
		OpsGet:   binary.BigEndian.Uint64(payload[0:]),
		OpsPut:   binary.BigEndian.Uint64(payload[8:]),
		P99GetUs: binary.BigEndian.Uint64(payload[16:]),
		P99PutUs: binary.BigEndian.Uint64(payload[24:]),
	}, nil
}

// rpc sends a command and returns (responseCmd, payload, error).
// If followRedirect is true and the server responds REDIRECT, the client
// reconnects to the new leader and retries (once).
func (c *Client) rpc(op cmd, payload []byte, followRedirect bool) (cmd, []byte, error) {
	c.mu.Lock()
	defer c.mu.Unlock()

	seq := c.seq.Add(1)
	hdr := buildHeader(op, uint32(len(payload)), seq)
	msg := append(hdr, payload...)

	if err := writeAll(c.conn, msg); err != nil {
		return 0, nil, fmt.Errorf("apexkv: send: %w", err)
	}

	rc, resp, err := readResponse(c.conn)
	if err != nil {
		return 0, nil, fmt.Errorf("apexkv: recv: %w", err)
	}

	if rc == cmdREDIRECT && followRedirect {
		leader := decodeStr(resp)
		if leader != "" && leader != "unknown" {
			c.addr = leader
			if err2 := c.reconnect(); err2 == nil {
				// Retry once on the new leader
				c.mu.Unlock()
				rc2, resp2, err3 := c.rpc(op, payload, false)
				c.mu.Lock()
				return rc2, resp2, err3
			}
		}
	}

	return rc, resp, nil
}

// ── Wire helpers ──────────────────────────────────────────────────────────────

func buildHeader(op cmd, payloadLen, seq uint32) []byte {
	hdr := make([]byte, hdrSize)
	binary.BigEndian.PutUint32(hdr[0:], wireMagic)
	hdr[4] = byte(op)
	hdr[5] = 0
	binary.BigEndian.PutUint16(hdr[6:], 0)
	binary.BigEndian.PutUint32(hdr[8:], payloadLen)
	binary.BigEndian.PutUint32(hdr[12:], seq)
	return hdr
}

func readResponse(conn net.Conn) (cmd, []byte, error) {
	hdr := make([]byte, hdrSize)
	if _, err := io.ReadFull(conn, hdr); err != nil {
		return 0, nil, err
	}
	magic := binary.BigEndian.Uint32(hdr[0:])
	if magic != wireMagic {
		return 0, nil, fmt.Errorf("bad magic 0x%08x", magic)
	}
	op      := cmd(hdr[4])
	plen    := binary.BigEndian.Uint32(hdr[8:])
	payload := make([]byte, plen)
	if plen > 0 {
		if _, err := io.ReadFull(conn, payload); err != nil {
			return 0, nil, err
		}
	}
	return op, payload, nil
}

func writeAll(conn net.Conn, data []byte) error {
	for len(data) > 0 {
		n, err := conn.Write(data)
		if err != nil {
			return err
		}
		data = data[n:]
	}
	return nil
}

func encodeStr(s string) []byte {
	b := []byte(s)
	out := make([]byte, 4+len(b))
	binary.BigEndian.PutUint32(out, uint32(len(b)))
	copy(out[4:], b)
	return out
}

func decodeStr(payload []byte) string {
	if len(payload) < 4 {
		return ""
	}
	n := binary.BigEndian.Uint32(payload)
	if int(n) > len(payload)-4 {
		return ""
	}
	return string(payload[4 : 4+n])
}

File: clients/rust/apex_kv.rs
// apex-kv Rust client
//
// Cargo.toml:
//   [dependencies]
//   apex-kv = { path = "." }
//   tokio = { version = "1", features = ["full"] }
//
// Example:
//   let mut client = ApexClient::connect("127.0.0.1:7001").await?;
//   client.put("hello", "world").await?;
//   let val = client.get("hello").await?;

use std::io::{self, Read, Write};
use std::net::{TcpStream, ToSocketAddrs};

const MAGIC: u32 = 0xA9ECBD10;
const HDR_SIZE: usize = 16;

#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq)]
enum Cmd {
    Get         = 0x01,
    Put         = 0x02,
    Del         = 0x03,
    Ping        = 0x04,
    Metrics     = 0x05,
    Ok          = 0x10,
    Value       = 0x11,
    NotFound    = 0x12,
    Error       = 0x13,
    Redirect    = 0x14,
    Pong        = 0x15,
    MetricsResp = 0x16,
    Unknown     = 0xFF,
}

impl From<u8> for Cmd {
    fn from(b: u8) -> Self {
        match b {
            0x01 => Cmd::Get,  0x02 => Cmd::Put, 0x03 => Cmd::Del,
            0x04 => Cmd::Ping, 0x05 => Cmd::Metrics,
            0x10 => Cmd::Ok,   0x11 => Cmd::Value, 0x12 => Cmd::NotFound,
            0x13 => Cmd::Error,0x14 => Cmd::Redirect, 0x15 => Cmd::Pong,
            0x16 => Cmd::MetricsResp,
            _ => Cmd::Unknown,
        }
    }
}

#[derive(Debug)]
pub enum ApexError {
    Io(io::Error),
    NotFound,
    NotLeader { leader: String },
    Server(String),
    Protocol(String),
}

impl std::fmt::Display for ApexError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ApexError::Io(e)               => write!(f, "io: {e}"),
            ApexError::NotFound            => write!(f, "key not found"),
            ApexError::NotLeader { leader} => write!(f, "not leader, redirect to {leader}"),
            ApexError::Server(s)           => write!(f, "server: {s}"),
            ApexError::Protocol(s)         => write!(f, "protocol: {s}"),
        }
    }
}

impl From<io::Error> for ApexError {
    fn from(e: io::Error) -> Self { ApexError::Io(e) }
}

pub type Result<T> = std::result::Result<T, ApexError>;

pub struct Metrics {
    pub ops_get:    u64,
    pub ops_put:    u64,
    pub p99_get_us: u64,
    pub p99_put_us: u64,
}

/// Synchronous APEX-KV client.
///
/// All methods are blocking. For async, wrap in `tokio::task::spawn_blocking`.
pub struct ApexClient {
    stream: TcpStream,
    seq: u32,
    addr: String,
    auto_redirect: bool,
}

impl ApexClient {
    /// Connect to an APEX-KV node.
    pub fn connect(addr: impl ToSocketAddrs + std::fmt::Display + Clone) -> Result<Self> {
        let stream = TcpStream::connect(addr.clone())?;
        stream.set_nodelay(true)?;
        Ok(Self {
            stream,
            seq: 0,
            addr: addr.to_string(),
            auto_redirect: true,
        })
    }

    /// Ping the node.
    pub fn ping(&mut self) -> Result<bool> {
        let (cmd, _) = self.rpc(Cmd::Ping, &[], false)?;
        Ok(cmd == Cmd::Pong)
    }

    /// Get a value. Returns `None` if not found.
    pub fn get(&mut self, key: &str) -> Result<Option<String>> {
        let payload = encode_str(key);
        let (cmd, resp) = self.rpc(Cmd::Get, &payload, false)?;
        match cmd {
            Cmd::Value    => Ok(Some(decode_str(&resp)?)),
            Cmd::NotFound => Ok(None),
            Cmd::Error    => Err(ApexError::Server(decode_str(&resp).unwrap_or_default())),
            _             => Err(ApexError::Protocol(format!("unexpected cmd {:?}", cmd))),
        }
    }

    /// Get a value with a default fallback.
    pub fn get_or_default(&mut self, key: &str, default: &str) -> Result<String> {
        Ok(self.get(key)?.unwrap_or_else(|| default.to_owned()))
    }

    /// Write a key-value pair. Follows leader redirects automatically.
    pub fn put(&mut self, key: &str, value: &str) -> Result<()> {
        let mut payload = encode_str(key);
        payload.extend(encode_str(value));
        let (cmd, resp) = self.rpc(Cmd::Put, &payload, true)?;
        if cmd == Cmd::Ok { return Ok(()); }
        Err(ApexError::Server(decode_str(&resp).unwrap_or_default()))
    }

    /// Delete a key. Returns true if key existed.
    pub fn del(&mut self, key: &str) -> Result<bool> {
        let payload = encode_str(key);
        let (cmd, _) = self.rpc(Cmd::Del, &payload, true)?;
        Ok(cmd == Cmd::Ok)
    }

    /// Fetch server metrics.
    pub fn metrics(&mut self) -> Result<Metrics> {
        let (cmd, resp) = self.rpc(Cmd::Metrics, &[], false)?;
        if cmd != Cmd::MetricsResp || resp.len() < 32 {
            return Err(ApexError::Protocol("bad metrics response".into()));
        }
        Ok(Metrics {
            ops_get:    u64::from_be_bytes(resp[0..8].try_into().unwrap()),
            ops_put:    u64::from_be_bytes(resp[8..16].try_into().unwrap()),
            p99_get_us: u64::from_be_bytes(resp[16..24].try_into().unwrap()),
            p99_put_us: u64::from_be_bytes(resp[24..32].try_into().unwrap()),
        })
    }

    // ── Internal ───────────────────────────────────────────────────────────────

    fn rpc(&mut self, cmd: Cmd, payload: &[u8], follow_redirect: bool) -> Result<(Cmd, Vec<u8>)> {
        self.seq = self.seq.wrapping_add(1);
        let hdr = build_header(cmd, payload.len() as u32, self.seq);
        self.stream.write_all(&hdr)?;
        self.stream.write_all(payload)?;

        let (rcmd, resp) = read_response(&mut self.stream)?;

        if rcmd == Cmd::Redirect && follow_redirect && self.auto_redirect {
            if let Ok(leader) = decode_str(&resp) {
                if !leader.is_empty() && leader != "unknown" {
                    self.addr = leader;
                    self.stream = TcpStream::connect(&self.addr)?;
                    self.stream.set_nodelay(true)?;
                    return self.rpc(cmd, payload, false);
                }
            }
        }

        Ok((rcmd, resp))
    }
}

fn build_header(cmd: Cmd, payload_len: u32, seq: u32) -> [u8; HDR_SIZE] {
    let mut hdr = [0u8; HDR_SIZE];
    hdr[0..4].copy_from_slice(&MAGIC.to_be_bytes());
    hdr[4] = cmd as u8;
    hdr[8..12].copy_from_slice(&payload_len.to_be_bytes());
    hdr[12..16].copy_from_slice(&seq.to_be_bytes());
    hdr
}

fn read_response(stream: &mut TcpStream) -> Result<(Cmd, Vec<u8>)> {
    let mut hdr = [0u8; HDR_SIZE];
    stream.read_exact(&mut hdr)?;
    let magic = u32::from_be_bytes(hdr[0..4].try_into().unwrap());
    if magic != MAGIC {
        return Err(ApexError::Protocol(format!("bad magic 0x{magic:08x}")));
    }
    let cmd  = Cmd::from(hdr[4]);
    let plen = u32::from_be_bytes(hdr[8..12].try_into().unwrap()) as usize;
    let mut payload = vec![0u8; plen];
    if plen > 0 {
        stream.read_exact(&mut payload)?;
    }
    Ok((cmd, payload))
}

fn encode_str(s: &str) -> Vec<u8> {
    let b = s.as_bytes();
    let mut out = (b.len() as u32).to_be_bytes().to_vec();
    out.extend_from_slice(b);
    out
}

fn decode_str(payload: &[u8]) -> Result<String> {
    if payload.len() < 4 {
        return Ok(String::new());
    }
    let n = u32::from_be_bytes(payload[0..4].try_into().unwrap()) as usize;
    if payload.len() < 4 + n {
        return Err(ApexError::Protocol("string truncated".into()));
    }
    String::from_utf8(payload[4..4+n].to_vec())
        .map_err(|_| ApexError::Protocol("invalid utf8".into()))
}

// ── Example main ──────────────────────────────────────────────────────────────
fn main() -> Result<()> {
    let addr = std::env::args().nth(1).unwrap_or_else(|| "127.0.0.1:7001".into());
    let mut client = ApexClient::connect(&addr)?;
    println!("Connected to {addr}  ping={}", client.ping()?);

    client.put("rust-key", "rust-value")?;
    match client.get("rust-key")? {
        Some(v) => println!("get rust-key = {v}"),
        None    => println!("not found"),
    }
    client.del("rust-key")?;

    let m = client.metrics()?;
    println!("metrics: ops_get={} ops_put={} p99_get={}µs",
             m.ops_get, m.ops_put, m.p99_get_us);
    Ok(())
}
